wire-ios-data-model/Source/Proteus/ProteusProvider.swift

```diff
@@ -0,0 +1,57 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public typealias KeyStorePerformBlock<T> = ((UserClientKeysStore) throws -> T)
+public typealias ProteusServicePerformBlock<T> = ((ProteusServiceInterface) throws -> T)
+
+public protocol ProteusProviding {
+
+    func perform<T>(
+        withProteusService proteusServiceBlock: ProteusServicePerformBlock<T>,
+        withKeyStore keyStoreBlock: KeyStorePerformBlock<T>
+    ) rethrows -> T
+
+}
+
+public class ProteusProvider: ProteusProviding {
+
+    private let context: NSManagedObjectContext
+
+    public init(context: NSManagedObjectContext) {
+        self.context = context
+    }
+
+    public func perform<T>(
+        withProteusService proteusServiceBlock: ProteusServicePerformBlock<T>,
+        withKeyStore keyStoreBlock: KeyStorePerformBlock<T>
+    ) rethrows -> T {
+
+        precondition(context.zm_isSyncContext, "ProteusProvider should only be used on the sync context")
+
+        if let proteusService = context.proteusService {
```

Comment:
The proteus service should only be available if the flag is true, so it doesn't seem necessary to check the flag. But it could be a good place to do it anyway as an extra verification

Reply:
I don't see where the proteusService is created...

Reply:
ProteusService is created here https://github.com/wireapp/wire-ios-mono/blob/0806881ea87daa2537f0017fd4820252c39c9a1f/wire-ios-sync-engine/Source/UserSession/ZMUserSession%2BCoreCrypto.swift#L57

Reply:
`ProteusProvider`'s responsibility is not to create the crypto service (`ProteusService` or `UserClientKeysStore`). It only "fetches" them from the managed object context when they're needed. They should be setup prior to using the `ProteusProvider`.

Maybe the name `ProteusProvider` is misleading